src/pipelines/step/composite/atomic.rs

```diff
@@ -0,0 +1,102 @@
+use {super::*, std::sync::Arc};
+
+pub struct AtomicMode;
+
+impl<P: Platform> CompositeStepMode<P> for AtomicMode {
+	async fn steps(
+		&self,
+		steps: &[Arc<StepInstance<P>>],
+		payload: Checkpoint<P>,
+		ctx: StepContext<P>,
+	) -> ControlFlow<P> {
+		let initial = payload.clone();
+		let mut current = payload;
+
+		for step in steps {
+			if ctx.deadline_reached() {
+				return ControlFlow::Break(initial);
+			}
+
+			match step.step(current, ctx.clone()).await {
+				ControlFlow::Ok(next) => current = next,
+				ControlFlow::Break(_) => return ControlFlow::Break(initial),
+				ControlFlow::Fail(error) => return ControlFlow::Fail(error),
+			}
+		}
+
+		if ctx.deadline_reached() {
+			ControlFlow::Break(initial)
+		} else {
+			ControlFlow::Ok(current)
+		}
+	}
+}
+
+#[macro_export]
+macro_rules! atomic {
+    ($($step:expr),+ $(,)?) => {{
+        let mut composite =
+            $crate::prelude::composite::CompositeStep::new(
+                $crate::prelude::composite::atomic::AtomicMode,
+            );
+        $(
+            composite.append_step($step);
+        )+
+        composite
+    }};
+}
```

Comment:
I love this macro! This could be an alternative to the syntax I proposed in the other PR.